### PR TITLE
format intermediate steps

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: libs/langchain/langchain/agents/format_scratchpad/log_to_messages.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from typing import List, Tuple, Union
+
+from langchain.schema.agent import AgentAction
+from langchain.schema.messages import AIMessage, BaseMessage, HumanMessage
+
+
+def format_log_to_messages(
+    intermediate_steps: List[Tuple[AgentAction, str]],
+    template_tool_response: str = "{observation}",
+) -> Union[str, List[BaseMessage]]:
```
Comment (nfcampos) [inline libs/langchain/langchain/agents/format_scratchpad/log_to_messages.py:10]: why this union return type?

File: libs/langchain/langchain/agents/format_scratchpad/log.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+def format_log_to_str(
+    intermediate_steps: List[Tuple[AgentAction, str]],
+    observation_prefix: str = "Observation: ",
+    llm_prefix: str = "Thought: ",
+) -> str:
+    """Construct the scratchpad that lets the agent continue its thought process."""
+    thoughts = ""
+    for action, observation in intermediate_steps:
+        thoughts += action.log
+        thoughts += f"\n{observation_prefix}{observation}\n{llm_prefix}"
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/format_scratchpad/log.py:15]: imo something like this makes it easier to see what's actually happening but up to you

```suggestion
    tmpl = f"""\
{{action.log}}
{observation_prefix}{{observation}}
{llm_prefix}"""
    for action, observation in intermediate_steps:
        thoughts += tmpl.format(action=action, observation=observation)
```

File: libs/langchain/langchain/agents/format_scratchpad/log_to_messages.py
Comment (baskaryan) [inline libs/langchain/langchain/agents/format_scratchpad/log_to_messages.py:1]: could this just live in format_scratchpad/log.py? given recent import speed considerations, wonder if we should generally start moving to fewer files?

File: libs/langchain/langchain/agents/format_scratchpad/openai_functions.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    if isinstance(agent_action, AgentActionMessageLog):
+        return list(agent_action.message_log) + [
+            _create_function_message(agent_action, observation)
+        ]
+    else:
+        return [AIMessage(content=agent_action.log)]
+
+
+def _create_function_message(
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/format_scratchpad/openai_functions.py:29]: nit: would call _convert_agent_action_to_function_message to match ^

File: libs/langchain/langchain/agents/format_scratchpad/openai_functions.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    """Format intermediate steps.
+    Args:
+        intermediate_steps: Steps the LLM has taken to date, along with observations
+    Returns:
+        list of messages to send to the LLM for the next prediction
+    """
+    messages = []
+
+    for intermediate_step in intermediate_steps:
+        agent_action, observation = intermediate_step
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/format_scratchpad/openai_functions.py:64]:
```suggestion
    for agent_action, observation in intermediate_steps:
```

File: libs/langchain/langchain/agents/format_scratchpad/xml.py
Comment (baskaryan) [inline libs/langchain/langchain/agents/format_scratchpad/xml.py:1]: file and function names confusing me a bit, should this be format_log_to_xml?

**Review by baskaryan — APPROVED** [review]

File: libs/langchain/langchain/agents/format_scratchpad/xml.py
Comment (hwchase17) [reply libs/langchain/langchain/agents/format_scratchpad/xml.py:1]: nope - cause this isnt working with the log, its working with the direct arguments

**Pull request merged**